src/hotspot/share/logging/logAsyncFlusher.cpp

```diff
@@ -0,0 +1,233 @@
+/*
+ * Copyright Amazon.com Inc. or its affiliates. All Rights Reserved.
+ * DO NOT ALTER OR REMOVE COPYRIGHT NOTICES OR THIS FILE HEADER.
+ *
+ * This code is free software; you can redistribute it and/or modify it
+ * under the terms of the GNU General Public License version 2 only, as
+ * published by the Free Software Foundation.
+ *
+ * This code is distributed in the hope that it will be useful, but WITHOUT
+ * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
+ * FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License
+ * version 2 for more details (a copy is included in the LICENSE file that
+ * accompanied this code).
+ *
+ * You should have received a copy of the GNU General Public License version
+ * 2 along with this work; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin St, Fifth Floor, Boston, MA 02110-1301 USA.
+ *
+ * Please contact Oracle, 500 Oracle Parkway, Redwood Shores, CA 94065 USA
+ * or visit www.oracle.com if you need additional information or have any
+ * questions.
+ *
+ */
+#include "precompiled.hpp"
+#include "jvm.h"
+#include "logging/logAsyncFlusher.hpp"
+#include "logging/logConfiguration.hpp"
+#include "logging/logFileOutput.hpp"
+#include "logging/logHandle.hpp"
+#include "runtime/atomic.hpp"
+
+void AsyncLogMessage::writeback() {
+  if (_message != NULL) {
+    _output.write_blocking(_decorations, _message);
+    os::free(_message);
+  }
+}
+
+void LogAsyncFlusher::enqueue_impl(const AsyncLogMessage& msg) {
+  assert_lock_strong(&_lock);
+
+  if (_buffer.size() >= _buffer_max_size)  {
```

Comment:
As mentioned in the CSR, I prefer dropping the newest message, not the oldest one. So there would be no reason to pop anything from the buffer.

Reply:
copy that.  I will update it.

Reply:
> One thing occurring to me was that a global buffer for all outputs is maybe not the best choice. One buffer per LogFileOutput seems like a cleaner, more natural fit:

That's another reason I introduce a hashmap for dropped counters.

I read you in the RFC thread about the "Global Big Lock" for all filed-based log outputs.  I know it's not scalable.I proposed to Volker splitting buffers for different outputs. Then the cost of synchronization would be effectively same as current UL. Volker countered that it wasn't an issue until users deploys many log files. it seems rare. We can leave it for future improvement.

Reply:
I think what Volker said makes sense, then `KVHashtable<LogFileOutput*, AsyncLogBuffer&>` downsized to 
`KVHashtable<LogFileOutput*, uint32_t>`...

Reply:
> > One thing occurring to me was that a global buffer for all outputs is maybe not the best choice. One buffer per LogFileOutput seems like a cleaner, more natural fit:
> 
> That's another reason I introduce a hashmap for dropped counters.
> 
> I read you in the RFC thread about the "Global Big Lock" for all filed-based log outputs. I know it's not scalable.I proposed to Volker splitting buffers for different outputs. Then the cost of synchronization would be effectively same as current UL. Volker countered that it wasn't an issue until users deploys many log files. it seems rare. We can leave it for future improvement.

Performance was not my main reason for suggesting this, but code simplicity.

Im fine with leaving this for a future RFE

Fior the record, here is the idea:
- keep async buffer, drop counter and async buffer lock as members in LogFileOutput
- have a global counter "enqueued-items"
- in LogFileOutput::write:
  - if enqueued-items=max, increase the drop counter in LogFileOutput
  - otherwise create buffer if needed, then enqueue into this buffer, increase global enqueued-items counter
- LogAsyncFlusher::flush becomes LogFileOutput::flush_async_buffer() or similar and:
  - prints its drop counter, if > 0
  - flushes its own buffer, same logic as today, but locks over its local _async_buffer_log
  - and decreases global enqueued-items counter
- flushing means iterating over all LogFileOutput objects and calling flush on them.


> I think what Volker said makes sense, then `KVHashtable<LogFileOutput*, AsyncLogBuffer&>` downsized to
> `KVHashtable<LogFileOutput*, uint32_t>`...

I'm loosing track here, where is Volker's remark and where is the first variant used?